dbmanager/dbmanager.h

```diff
@@ -0,0 +1,66 @@
+#include <stdio.h>
+#include <stdlib.h>
+
+
+/*  datatype of each node in the database.
+ *  @key - a unique integer value.
+ *  @val - pointer to the data.
+ *  @left, @right - pointers to children.
+ */
+typedef struct dbNode {
+  int key;
+  void* val;
+  struct dbNode* left;
+  struct dbNode* right;
+} dbNode;
+
+/*
+ *  keeps track of head of database.
+ *  supports only one database.
+ *  support for multiple databases can be extended by storing a list of heads.
+ */
+dbNode* dbHead = NULL;
+
+/*
+ *  initializes database.
+ *  can only be called once for signle database application.
+ *  @return - 0 if success, -1 if failure.
+ */
+
+int dbInit();
```

Comment:
If you instead return a context from the init function you dont need to make it a singleton and the lib becomes more flexible.
define the type DB; typedef struct DB; and return DB* from dbInit (youll define the meat of the struct within the source component). You will also need to add this context to all the functions though; ex. int dbAdd(DB*, int key, void* val) etc.
The DB* could be the head but you might want to include other non tree data in the context so would recommend the DB struct including the head instead.